llvm/lib/Analysis/LazyValueInfo.cpp

```diff
@@ -741,27 +986,32 @@ LazyValueInfoImpl::solveBlockValueNonLocal(Value *Val, BasicBlock *BB) {
   std::optional<BBLatticeElementMap> PredLatticeElements;
   if (PerPredRanges)
     PredLatticeElements = std::make_optional<BBLatticeElementMap>();
-  for (BasicBlock *Pred : predecessors(BB)) {
-    // Skip self loops.
-    if (Pred == BB)
-      continue;
-    std::optional<ValueLatticeElement> EdgeResult = getEdgeValue(Val, Pred, BB);
-    if (!EdgeResult)
-      // Explore that input, then return here
-      return std::nullopt;
 
-    Result.mergeIn(*EdgeResult);
+  {
+    ValueLatticeMerger Merger(Result);
+    for (BasicBlock *Pred : predecessors(BB)) {
+      // Skip self loops.
+      if (Pred == BB)
+        continue;
+      std::optional<ValueLatticeElement> EdgeResult =
+          getEdgeValue(Val, Pred, BB);
+      if (!EdgeResult)
+        // Explore that input, then return here
+        return std::nullopt;
+
+      Merger.mergeIn(*EdgeResult);
```

Comment:
Is it necessary to manage to extend this change to SCCP as well?
The monotonicity of the merge function is a sufficient, but not a necessary, condition for the convergence of fixpoint iteration.